keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,888 @@
+
+## Mutating Buckets
+
+As of the current design of COSI, mutating bucket properties is not supported. However, the current design does not prevent us from supporting it in the future. Mutable properties include encryption, object lifecycle configuration, replication configuration etc. These properties will be supported in future versions along with the capability to mutate them.
+
+These properties will be specified in the BucketRequest and follow the same pattern of events as Bucket creation. i.e. Bucket API object will be updated to reflect the properties set in BucketRequest, and then a controller will pick-up these changes and call the appropriate APIs to reflect them in the backend.
+
+# Object Lifecycle
+
+The following resources are managed by admins
+
+- BucketClass
+- BucketAccessClass
+
+The following resources are managed by a user. These are created with a reference to their corresponding class objects:
+
+- BucketClaim -> BucketClass
+- BucketAccess -> Bucket, BucketClaim, BucketAccessClass
+
+The COSI controller responds by creating the following objects
+
+- BucketRequest <-> Bucket (1:1)
+- BucketClaim -> Bucket
+
+Notes:
+
+ - There are **NO** cycles in the relationship graph of the above mentioned API objects.
+ - Mutations are not supported in the API.
+
+When a user deletes the BucketRequest, then depending on the DeletionPolicy, the following happens:
+
+- If deletionPolicy is Delete, then Bucket deletion is also triggered. 
+- If deletionPolicy is Retain, then Bucket is left as is, but the BucketRequest is deleted.
 
-message ProvisionerGetInfoResponse {
-  // The name MUST follow domain name notation format
-  // (https://tools.ietf.org/html/rfc1035#section-2.3.1). It SHOULD
-  // include the plugin's host company name and the plugin name,
-  // to minimize the possibility of collisions. It MUST be 63
-  // characters or less, beginning and ending with an alphanumeric
-  // character ([a-z0-9A-Z]) with dashes (-), dots (.), and
-  // alphanumerics between. This field is REQUIRED.
-  string name = 1;
+Only when all accessors (BucketAccesses) of the Bucket are deleted, is the Bucket itself cleaned up. Having orphaned buckets in the cluster is an invalid state, unless the Bucket was imported into the cluster. 
+
+When a user deletes a BucketAccess, the corresponding secret/serviceaccount are also deleted. If a pod has that secret mounted when delete is called, then the secret will not be deleted, but will instead have its deletionTimestamp set. In this way, access to a Bucket is preserved until the application pod dies. 
+
+When an admin deletes any of the class objects, it does not affect existing Buckets as fields from the class objects are copied into the Buckets during creation. 
+
+If a Bucket is manually deleted by an admin, without deleting the corresponding BucketClaim, then the cluster is considered to be in an invalid state. Manual recovery is possible if data is not already lost. 
```

Comment:
This makes sense, since we also considered adding a finalizer on the Bucket for the BucketClaim. This way, the Bucket cannot be deleted before BucketClaim is deleted. I've made a note of it in the KEP